word_language_model/model.py

```diff
@@ -1,57 +1,165 @@
+        if self.model_type == 'Transformer':
+            self._reset_parameters()
+
+        else:
+            # Optionally tie weights as in:
+            # "Using the Output Embedding to Improve Language Models" (Press & Wolf 2016)
+            # https://arxiv.org/abs/1608.05859
+            # and
+            # "Tying Word Vectors and Word Classifiers: A Loss Framework for Language Modeling" (Inan et al. 2016)
+            # https://arxiv.org/abs/1611.01462
+            if tie_weights:
+                if nhid != ninp:
+                    raise ValueError('When using the tied flag, nhid must be equal to emsize')
+                self.decoder.weight = self.encoder.weight
+
+            self.init_weights()
 
-        self.init_weights()
+            self.model_type = model_type
+            self.nhid = nhid
+            self.nlayers = nlayers
 
-        self.rnn_type = rnn_type
-        self.nhid = nhid
-        self.nlayers = nlayers
+    def _reset_parameters(self):
+        r"""Initiate parameters in the model."""
+        assert self.model_type == 'Transformer'
+
+        for p in self.parameters():
+            if p.dim() > 1:
+                nn.init.xavier_uniform_(p)
+
+    def _set_src_mask(self, mask):
+        # self.src_mask: [source sequence length, source sequence length]
+        assert self.model_type == 'Transformer'
+        self.src_mask = mask
+
+    def _generate_square_subsequent_mask(self, sz):
+        assert self.model_type == 'Transformer'
+        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
+        mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+        return mask
 
     def init_weights(self):
+        assert self.model_type != 'Transformer'
         initrange = 0.1
         self.encoder.weight.data.uniform_(-initrange, initrange)
         self.decoder.bias.data.zero_()
         self.decoder.weight.data.uniform_(-initrange, initrange)
 
-    def forward(self, input, hidden):
-        emb = self.drop(self.encoder(input))
-        output, hidden = self.rnn(emb, hidden)
-        output = self.drop(output)
-        decoded = self.decoder(output.view(output.size(0)*output.size(1), output.size(2)))
-        return decoded.view(output.size(0), output.size(1), decoded.size(1)), hidden
+    def forward(self, input, hidden, has_mask=True):
+        if self.model_type == 'Transformer':
+            if has_mask:
+                device = input.device
+                if self.src_mask is None or self.src_mask.size(0) != len(input):
+                    mask = self._generate_square_subsequent_mask(len(input)).to(device)
+                    self._set_src_mask(mask)
+            else:
+                self._set_src_mask(None)
+
+            emb = self.drop(self.encoder(input))
+            output = emb * math.sqrt(self.ninp)
+            output = self.pos_encoder(output)
+            output = self.model(output, self.src_mask)
```

Comment:
I'm confused about why we still need this mask, if the decoder is a Linear layer?